msal/managed_identity.py

```diff
@@ -0,0 +1,528 @@
+
+    @classmethod
+    def is_user_assigned(cls, unknown):
+        return isinstance(unknown, UserAssignedManagedIdentity) or (
+            isinstance(unknown, dict)
+            and unknown.get(cls.ID_TYPE) in cls._types_mapping
+            and unknown.get(cls.ID))
+
+    def __init__(self, identifier=None, id_type=None):
+        # Undocumented. Use subclasses instead.
+        super(ManagedIdentity, self).__init__({
+            self.ID_TYPE: id_type,
+            self.ID: identifier,
+        })
+
+
+class SystemAssignedManagedIdentity(ManagedIdentity):
+    """Represent a system-assigned managed identity.
+
+    It is equivalent to a Python dict of::
+
+        {"ManagedIdentityIdType": "SystemAssigned", "Id": None}
+
+    or a JSON blob of::
+
+        {"ManagedIdentityIdType": "SystemAssigned", "Id": null}
+    """
+    def __init__(self):
+        super(SystemAssignedManagedIdentity, self).__init__(id_type=self.SYSTEM_ASSIGNED)
+
+
+class UserAssignedManagedIdentity(ManagedIdentity):
+    """Represent a user-assigned managed identity.
+
+    Depends on the id you provided, the outcome is equivalent to one of the below::
+
+        {"ManagedIdentityIdType": "ClientId", "Id": "foo"}
+        {"ManagedIdentityIdType": "ResourceId", "Id": "foo"}
+        {"ManagedIdentityIdType": "ObjectId", "Id": "foo"}
+    """
+    def __init__(self, client_id=None, resource_id=None, object_id=None):
+        if client_id and not resource_id and not object_id:
+            super(UserAssignedManagedIdentity, self).__init__(
+                id_type=self.CLIENT_ID, identifier=client_id)
+        elif not client_id and resource_id and not object_id:
+            super(UserAssignedManagedIdentity, self).__init__(
+                id_type=self.RESOURCE_ID, identifier=resource_id)
+        elif not client_id and not resource_id and object_id:
+            super(UserAssignedManagedIdentity, self).__init__(
+                id_type=self.OBJECT_ID, identifier=object_id)
+        else:
+            raise ValueError(
+                "You shall specify one of the three parameters: "
+                "client_id, resource_id, object_id")
+
+
+class _ThrottledHttpClient(ThrottledHttpClientBase):
+    def __init__(self, http_client, http_cache):
+        super(_ThrottledHttpClient, self).__init__(http_client, http_cache)
+        self.get = IndividualCache(  # All MIs (except Cloud Shell) use GETs
+            mapping=self._expiring_mapping,
+            key_maker=lambda func, args, kwargs: "POST {} hash={} 429/5xx/Retry-After".format(
+                args[0],  # It is the endpoint, typically a constant per MI type
+                _hash(
+                    # Managed Identity flavors have inconsistent parameters.
+                    # We simply choose to hash them all.
+                    str(kwargs.get("params")) + str(kwargs.get("data"))),
+                ),
+            expires_in=_parse_http_429_5xx_retry_after,
```

Comment:
I don't remember the 5xx rule, but please ensure the throttling timeout is not bigger than a few seconds.

Reply:
FYI: We are reusing the existing throttling logic defined in MSAL. The actual timeout time is largely controlled by the `Retry-After` header which is determined by the MI service. So, we shall be good here.

Reply:
Sure, but in case of 5xx you default to some constant delay. I remember that the constant was set to 1-2 minutes, which might make sense for public client, but it's unacceptable for CCA.

Reply:
You made a good point about throttling timeout for CCA could have been different than PCA.

Would you also suggest us to change the existing throttling behavior inside MSAL's non-MSI code path? It turns out MSAL Python's PCA and CCA share same default throttling timeout (which is 60 seconds) for years now. :-/

Reply:
Yeah, same for MSAL .NET :(. But I think MSAL.NET will be able to interally retry without hitting the throttling.